Add `clipboard_has/get_image` methods to DisplayServer

deakcor opened the pull request:
![pr_paste_image_preview](https://user-images.githubusercontent.com/40196601/182387754-d75f7806-6d75-4038-90fc-7220f900eafd.gif)

Following proposal [#2949](https://github.com/godotengine/godot-proposals/issues/2949#issue-937370011), added

`clipboard_get_image()` to get image from the clipboard

`clipboard_has_image()` to check if the clipboard has an image. It's faster than `clipboard_get_image()` so It may be useful to make a check (to disable a "paste" button for example)

Override of `clipboard_has()`, may be faster for a huge text.

### They are only implemented on Windows and MacOS. Should be added on x11 (couldn't get a working result).

Here a code to try quickly (just add a texturerect in child) :

```
if DisplayServer.clipboard_has():
	print("clipboard txt : ",DisplayServer.clipboard_get())
if DisplayServer.clipboard_has_image():
	print("clipboard img : ",DisplayServer.clipboard_get_image().get_data())
	$TextureRect.texture=ImageTexture.create_from_image(DisplayServer.clipboard_get_image())
```

Also should be nice to rename `clipboard_has()` to  `clipboard_has_text()` and  `clipboard_get()` to  `clipboard_get_text()`. Or make `Variant` as result of  `clipboard_get()` and `clipboard_has()` should add an optional param to indicate what type it should check.

_Production edit:_ Closes https://github.com/godotengine/godot-proposals/issues/2949.

deakcor committed:
add clipboard get image methods for windows and macos

Co-Authored-By: RedworkDE <10944644+RedworkDE@users.noreply.github.com>
Co-Authored-By: bruvzg <7645683+bruvzg@users.noreply.github.com>

akien-mga commented:
I think that's pretty useful to support. We'll indeed need a X11 implementation too (could be in a follow-up PR if someone else wants to work on it), and assess other platforms (Android, iOS, Web).

> Also should be nice to rename `clipboard_has()` to `clipboard_has_text()` and `clipboard_get()` to `clipboard_get_text()`. 

I agree, this should be done if this PR is accepted. That would also solve the awkwardness of `clipboard_has()` and `clipboard_get()` Yoda-speak, because then it would be more obvious that all those methods are `clipboard_<method>`.

There should also be support for `clipboard_set_image()` IMO, and `clipboard_set()` should be `clipboard_set_text()`.
`clipboard_set_image()` could be quite convenient to let users grab a screenshot for easy sharing on social media / Discord, etc.

deakcor commented:
I began to replace "clipboard" methods with "text" at the end. And I saw `clipboard_set_primary` and `clipboard_get_primary` methods, it duplicates a bit all the methods only for x11 system. I don't know if it exists for the images the primary ? But if yes it can become a bit messy.

Shouldn't be easier to add this in `clipboard_set_text` method ? Like `clipboard_set_text(text,primary=false)`. So that way you can do also `clipboard_has_text(primary=false)` without new methods.

m4gr3d commented:
> I think that's pretty useful to support. We'll indeed need a X11 implementation too (could be in a follow-up PR if someone else wants to work on it), and assess other platforms (Android, iOS, Web).

Android provides the ability to [copy/paste complex data structure](https://developer.android.com/guide/topics/text/copy-paste#Provider), so this can be added as well as a follow-up PR once the initial one has landed.

fritol commented:
wen in the in the main?

Calinou commented:
> wen in the in the main?

4.0 is in feature freeze, so this will have to wait for a future 4.x release to be merged. Also, this pull request needs to be rebased against the latest `master` branch first.

Steveplays28 commented:
Is there anything I could help with to get this merged?
Would be a really nice feature for making apps.

Calinou commented:
> Is there anything I could help with to get this merged? Would be a really nice feature for making apps.

4.1 is currently in feature freeze, but you can [test this PR locally](https://docs.godotengine.org/en/latest/contributing/workflow/testing_pull_requests.html) and make sure it works as expected. We also recommend creating a publicly accessible testing project (cc @deakcor). That said, this PR needs to be rebased against the latest `master` branch before it can be merged in a future 4.x release.

> I began to replace "clipboard" methods with "text" at the end. And I saw clipboard_set_primary and clipboard_get_primary methods, it duplicates a bit all the methods only for x11 system. I don't know if it exists for the images the primary ? But if yes it can become a bit messy.

Note that this particular part of the PR needs to be reverted, as we can't break compatibility anymore now that 4.0 is released.

deakcor commented:
> > Is there anything I could help with to get this merged? Would be a really nice feature for making apps.
> 
> 4.1 is currently in feature freeze, but you can [test this PR locally](https://docs.godotengine.org/en/latest/contributing/workflow/testing_pull_requests.html) and make sure it works as expected. We also recommend creating a publicly accessible testing project (cc @deakcor if they still have a copy of the project). That said, this PR needs to be rebased against the latest `master` branch before it can be merged in a future 4.x release.
> 
> > I began to replace "clipboard" methods with "text" at the end. And I saw clipboard_set_primary and clipboard_get_primary methods, it duplicates a bit all the methods only for x11 system. I don't know if it exists for the images the primary ? But if yes it can become a bit messy.
> 
> Note that this particular part of the PR needs to be reverted, as we can't break compatibility anymore now that 4.0 is released.

So I rebased and reverted the part with "_text" behind clipboard methods. If you validate the changes, let me know and I'll squash the commits into one

akien-mga commented:
Looks good to me overall. The classref needs to be updated, see the diff in https://github.com/godotengine/godot/actions/runs/5513089283/jobs/10050780703?pr=63826

CC @bruvzg to review Windows and macOS implementations.

bruvzg commented on platform/macos/display_server_macos.mm:
```diff
@@ -2100,6 +2101,37 @@
+	NSBitmapImageRep *bitmap = [NSBitmapImageRep imageRepWithData:data];
+	NSData *pngData = [bitmap representationUsingType:NSPNGFileType properties:@{}];
+	image.instantiate();
+	UInt8 buf[pngData.length];
+	[pngData getBytes:buf length:pngData.length];
+	PNGDriverCommon::png_to_image(buf, pngData.length, false, image);
```
```suggestion
	PNGDriverCommon::png_to_image((const uint8_t *)pngData.bytes, pngData.length, false, image);
```

bruvzg commented on platform/macos/display_server_macos.mm:
```diff
@@ -2100,6 +2101,37 @@
+	if (!result)
+		return image;
+	NSData *data = [pasteboard dataForType:result];
+	if (!data)
+		return image;
```
```suggestion
	if (!data) {
		return image;
	}
```

deakcor commented on platform/windows/display_server_windows.cpp:
```diff
@@ -341,6 +342,67 @@ String DisplayServerWindows::clipboard_get() const {
+				GlobalUnlock(mem);
+			}
+		}
+	}
```
I think it should not happen, I see it has not be done in clipboard_get() method too, maybe returning empty image is enough ? 
Btw the CF_DIB format is the format used for screenshots but the way to handle data is very handmaded so it's really not optimized, I think it can be way more optimized.

deakcor commented:
Strange, the checks succeeded before I squashed the commits but normally there isn't any issue. Let me know if I need to change anything.

bruvzg commented:
> Strange, the checks succeeded before I squashed the commits but normally there isn't any issue. Let me know if I need to change anything.

It's not related to PR and caused by https://github.com/godotengine/godot/issues/78749, I have restarted failed job.

RedworkDE commented on servers/display_server.cpp:
```diff
@@ -362,10 +362,18 @@ String DisplayServer::clipboard_get() const {
 }
 
+bool DisplayServer::clipboard_has_image() const {
+	return !clipboard_get_image().is_valid();
```
```suggestion
	return clipboard_get_image().is_valid();
```

RedworkDE commented on platform/windows/display_server_windows.cpp:
```diff
@@ -341,6 +342,67 @@ String DisplayServerWindows::clipboard_get() const {
+}
+
+bool DisplayServerWindows::clipboard_has_image() const {
+	return IsClipboardFormatAvailable(CF_DIB);
```
This should also check for PNG or there will be discrepancies if the clipboard only contains PNG data and not DIB

deakcor replied:
I'll double check but CF_DIB is normally always available

RedworkDE commented on platform/macos/display_server_macos.mm:
```diff
@@ -2100,6 +2101,37 @@
+Ref<Image> DisplayServerMacOS::clipboard_get_image() const {
+	Ref<Image> image;
+	NSPasteboard *pasteboard = [NSPasteboard generalPasteboard];
+	NSString *result = [pasteboard availableTypeFromArray:[NSArray arrayWithObjects:NSPasteboardTypeTIFF, NSPasteboardTypePNG, nil]];
```
`NSPasteboardTypeTIFF` can this cause issues with the `representationUsingType:NSPNGFileType` below?

bruvzg replied:
`NSBitmapImageRep::representationUsingType` should encode it to the selected type regardless of source type.

YuriSizov commented:
Thanks! This is great! No we just need a volunteer to implement this for Linux :)
